### PR TITLE
Time zone related be function

### PR DESCRIPTION
**Pull request by HangyuanLiu**
Time zone related be function
#1583

### CODE REVIEW

File: be/src/runtime/timestamp_value.h [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+           << std::setw(2) << std::setfill('0') << ret_ptime.date().day() << " "
+           << std::setw(2) << std::setfill('0') << ret_ptime.time_of_day().hours() << ":"
+           << std::setw(2) << std::setfill('0') << ret_ptime.time_of_day().minutes() << ":"
+           << std::setw(2) << std::setfill('0') << ret_ptime.time_of_day().seconds();
+        return ss.str();
+    }
+
+    void to_datetime_val(doris_udf::DateTimeVal *tv) const {
+        boost::posix_time::ptime p = boost::posix_time::from_time_t(val / 1000);
+        int _year = p.date().year();
```
Comment (morningman) [inline be/src/runtime/timestamp_value.h]: Your code style is not compatible with Doris C++ code style.
private member should starts with `_`.
local variables should not start with `_`

**Review by morningman — CHANGES_REQUESTED** [review]

File: be/src/runtime/timestamp_value.h [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+class TimestampValue {
+public:
+    TimestampValue() { }
+
+    TimestampValue(int64_t timestamp) {
+        val = timestamp;
+    }
+
+    bool from_date_time_value(const DateTimeValue& tv, std::string timezone) {
```
Comment (imay) [inline be/src/runtime/timestamp_value.h]: You should add some comments for these functions

File: be/src/runtime/timestamp_value.h [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    static boost::local_time::time_zone_ptr find_timezone(const std::string &tz);
+
+private:
+    static const char *_s_timezone_database_str;
+    static boost::local_time::tz_database _s_tz_database;
+};
+
+class TimestampValue {
```
Comment (imay) [inline be/src/runtime/timestamp_value.h]: I think we just implement a constructor with (timestamp and timezone) in DataTimeValue is enough. And we can reuse code of DateTimeValue

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/testutil/function_utils.cpp [hunk 4]
```diff
@@ -34,6 +34,15 @@ FunctionUtils::FunctionUtils() {
     _fn_ctx = FunctionContextImpl::create_context(
         _state, _memory_pool, return_type, arg_types, 0, false);
 }
+FunctionUtils::FunctionUtils(RuntimeState* state) {       
+        _state = state;
```
Comment (imay) [inline be/src/testutil/function_utils.cpp]: 4 spaces indent

File: be/src/runtime/runtime_state.h [hunk 5]
```diff
@@ -136,8 +136,11 @@ class RuntimeState {
     int num_scanner_threads() const {
         return _query_options.num_scanner_threads;
     }
-    const DateTimeValue* now() const {
-        return _now.get();
+    int64_t timestamp() const {
+        return _timestamp;
+    }
+    const std::string timezone() const {
```
Comment (imay) [inline be/src/runtime/runtime_state.h]:
```suggestion
    const std::string& timezone() const {
```

File: be/src/runtime/datetime_value.cpp [hunk 6]
```diff
@@ -1524,43 +1524,42 @@ bool DateTimeValue::date_add_interval(const TimeInterval& interval, TimeUnit uni
...
-    }
-    return seconds;
-}
-
-bool DateTimeValue::from_unixtime(int64_t seconds) {
-    if (seconds < 0) {
+int64_t DateTimeValue::unix_timestamp(const std::string& timezone) const {
+    boost::local_time::time_zone_ptr local_time_zone = TimezoneDatabase::find_timezone(timezone);
+    if (local_time_zone == nullptr) {
         return false;
```
Comment (imay) [inline be/src/runtime/datetime_value.cpp]: unix_timestamp's return type is int64_t, it is better not to return `false`

File: be/src/runtime/datetime_value.h [hunk 7]
```diff
@@ -435,7 +436,8 @@ class DateTimeValue {
     }
 
     int64_t second_diff(const DateTimeValue& rhs) const {
-        return unix_timestamp() - rhs.unix_timestamp();
+        return unix_timestamp(TimezoneDatabase::default_time_zone) 
+            - rhs.unix_timestamp(TimezoneDatabase::default_time_zone);
```
Comment (imay) [inline be/src/runtime/datetime_value.h]: For second_diff which is a literal compare, I think we should not introduce timezone. Which can bring some cost which isn't needed. Such as search if timezone exists.

And I think we can introduce another seconds such as seconds from 0000-01-01 00:00:00, which has no relation with time zone.

File: be/src/exprs/timestamp_functions.cpp [hunk 8]
```diff
@@ -470,7 +372,8 @@ DoubleVal TimestampFunctions::time_diff(
     }
     const DateTimeValue& ts_value1 = DateTimeValue::from_datetime_val(ts_val1);
     const DateTimeValue& ts_value2 = DateTimeValue::from_datetime_val(ts_val2);
-    int64_t timediff = ts_value1.unix_timestamp() - ts_value2.unix_timestamp();
+    int64_t timediff = ts_value1.unix_timestamp(ctx->impl()->state()->timezone()) 
+        - ts_value2.unix_timestamp(ctx->impl()->state()->timezone());
```
Comment (imay) [inline be/src/exprs/timestamp_functions.cpp]: use another method to do this?

File: be/src/exprs/timezone_db.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+namespace doris {
+
+class TimezoneDatabase {
+public:
+    TimezoneDatabase();
+
+    ~TimezoneDatabase();
+
+    static void init() {
+        TimezoneDatabase();
```
Comment (imay) [inline be/src/exprs/timezone_db.h]: what this mean?

I think what you need is not this TimezoneDatabase constructor. 
You can move `TimezoneDatabase()` to this `init()`. 

I think there is no need to have object of this class. we only just need some static functions

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/runtime/datetime_value.h [hunk 10]
```diff
@@ -332,9 +333,9 @@ class DateTimeValue {
     // Add interval 
     bool date_add_interval(const TimeInterval& interval, TimeUnit unit);
 
-    int unix_timestamp() const;
+    int64_t unix_timestamp(const std::string& timezone) const;
```
Comment (imay) [inline be/src/runtime/datetime_value.h]: add some comments for these two functions

File: be/src/exprs/timestamp_functions.cpp [hunk 11]
```diff
@@ -485,142 +385,112 @@ IntVal TimestampFunctions::date_diff(
...
+    dtv.to_datetime_val(&return_val);
+    return return_val;
+}
 
-        if (tzp->dst_zone_abbrev() == tz) {
-            return tzp;
-        }
+DateTimeVal TimestampFunctions::now(FunctionContext* context) {
+    DateTimeValue dtv;
+    dtv.from_unixtime(context->impl()->state()->timestamp() / 1000,
```
Comment (imay) [inline be/src/exprs/timestamp_functions.cpp]: check this function's return?

File: be/src/exprs/timestamp_functions.cpp [hunk 12]
```diff
@@ -485,142 +385,112 @@ IntVal TimestampFunctions::date_diff(
...
+    dtv.to_datetime_val(&return_val);
+    return return_val;
+}
 
-        if (tzp->dst_zone_name() == tz) {
-            return tzp;
-        }
+DoubleVal TimestampFunctions::curtime(FunctionContext* context) {
+    DateTimeValue dtv;
+    dtv.from_unixtime(context->impl()->state()->timestamp() / 1000,
```
Comment (imay) [inline be/src/exprs/timestamp_functions.cpp]: check return value?

File: be/src/runtime/datetime_value.cpp [hunk 13]
```diff
@@ -1524,43 +1524,42 @@ bool DateTimeValue::date_add_interval(const TimeInterval& interval, TimeUnit uni
     return true;
 }
 
-int DateTimeValue::unix_timestamp() const {
-    int64_t days = daynr() - calc_daynr(1970, 1, 1);
-    if (days < 0) {
+int64_t DateTimeValue::unix_timestamp(const std::string& timezone) const {
```
Comment (imay) [inline be/src/runtime/datetime_value.cpp]: I think you should return a bool value if timezone is not valid

File: be/src/runtime/datetime_value.h [hunk 14]
```diff
@@ -331,10 +332,14 @@ class DateTimeValue {
...
     // Add interval 
     bool date_add_interval(const TimeInterval& interval, TimeUnit unit);
-
-    int unix_timestamp() const;
-
-    bool from_unixtime(int64_t);
+    
+    //unix_timestamp is called with a timezone argument,
+    //it returns seconds of the value of date literal since '1970-01-01 00:00:00' UTC
+    bool unix_timestamp(int64_t& timestamp, const std::string& timezone) const;
```
Comment (imay) [inline be/src/runtime/datetime_value.h]:
```suggestion
    bool to_unix_timestamp(int64_t* timestamp, const std::string& timezone) const;
```

Comment (HangyuanLiu) [reply be/src/runtime/datetime_value.h]: ok

**Review by imay — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**